extensions/sql-migration/src/models/loginMigrationModel.ts

```diff
@@ -0,0 +1,247 @@
+	switch (step) {
+		case LoginMigrationStep.NotStarted:
+			return constants.NOT_STARTED;
+		case LoginMigrationStep.MigrateLogins:
+			return constants.MIGRATE_LOGINS;
+		case LoginMigrationStep.EstablishUserMapping:
+			return constants.ESTABLISH_USER_MAPPINGS;
+		case LoginMigrationStep.MigrateServerRolesAndSetPermissions:
+			return constants.MIGRATE_SERVER_ROLES_AND_SET_PERMISSIONS;
+		case LoginMigrationStep.MigrationCompleted:
+			return constants.LOGIN_MIGRATION_COMPLETED;
+		default:
+			return ""
+	}
+}
+
+export enum LoginMigrationState {
+	Pending = 'Pending',
+	Running = 'Running',
+	Succeeded = 'Succeeded',
+	Failed = 'Failed',
+	Canceled = 'Canceled',
+}
+
+export const LoginMigrationStateToMultiStepState: constants.LookupTable<MultiStepState> = {
+	[LoginMigrationState.Pending]: MultiStepState.Pending,
+	[LoginMigrationState.Running]: MultiStepState.Running,
+	[LoginMigrationState.Succeeded]: MultiStepState.Succeeded,
+	[LoginMigrationState.Failed]: MultiStepState.Failed,
+	[LoginMigrationState.Canceled]: MultiStepState.Canceled,
+	default: MultiStepState.Pending
+};
+
+
+export interface LoginMigrationStepState {
+	loginName: string;
+	stepName: LoginMigrationStep;
+	status: LoginMigrationState;
+	errors: string[];
+}
+
+export interface Login {
+	loginName: string;
+	overallStatus: LoginMigrationState;
+	statusPerStep: Map<LoginMigrationStep, LoginMigrationStepState>;
+}
+
+export class LoginMigrationModel {
+	public resultsPerStep: Map<mssql.LoginMigrationStep, mssql.StartLoginMigrationResult>;
+	public collectedSourceLogins: boolean = false;
+	public collectedTargetLogins: boolean = false;;
+	public loginsOnSource: LoginTableInfo[] = [];
+	public loginsOnTarget: string[] = [];
+	private _currentStepIdx: number = 0;;
+	private _logins: Map<string, Login>;
+	private _loginMigrationSteps: LoginMigrationStep[] = [];
+
+	constructor() {
+		this.resultsPerStep = new Map<mssql.LoginMigrationStep, mssql.StartLoginMigrationResult>();
+		this._logins = new Map<string, Login>();
+		this.SetLoginMigrationSteps();
+	}
+
+	public get currentStep(): LoginMigrationStep {
+		return this._currentStepIdx >= this._loginMigrationSteps.length ? LoginMigrationStep.MigrationCompleted : this._loginMigrationSteps[this._currentStepIdx];
+	}
+
+	public get isMigrationComplete(): boolean {
+		return this._currentStepIdx === this._loginMigrationSteps.length;
+	}
+
+	public AddLoginMigrationResults(step: LoginMigrationStep, newResult: mssql.StartLoginMigrationResult): void {
+		const exceptionMap = this._getExceptionMapWithNormalizedKeys(newResult.exceptionMap);
+		this._currentStepIdx = this._loginMigrationSteps.findIndex(s => s === step) + 1;
+
+		for (const loginName of this._logins.keys()) {
+			const status = loginName in exceptionMap ? LoginMigrationState.Failed : LoginMigrationState.Succeeded;
+			let errors = loginName in exceptionMap ? this._extractErrors(exceptionMap, loginName) : [];
```

Comment:
nit:  prefer 'const' over 'let' when possible.

Reply:
agreed. I originally had it and had to change it to add the inner exception messages. I ended up cleaning it up with a private member function to get all the errors and forgot to change it back to const.